daemon/policy.go

```diff
@@ -302,6 +312,86 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 	return newRev, nil
 }
 
+type policyAddQueueOptions struct {
+	rules   policyAPI.Rules
+	opts    *AddOptions
+	context *policyQueueContext
+	reFetch bool
+}
+
+type policyQueueContext struct {
+	policyRevision uint64
+	err            error
+	done           chan bool
+	totalTime      spanstat.SpanStat
+	queueTime      spanstat.SpanStat
+}
+
+// policyAddToQueue adds the request to policyAdd in the queue and return a
+// context that will be updated when the done channel is closed.
+func (d *Daemon) policyAddToQueue(rules policyAPI.Rules, opts *AddOptions, reFetch bool) *policyQueueContext {
```

Comment:
The memory sharing of the returned value seems a little unusual here, basically the result is created and then passed to another goroutine through a channel, then the other goroutine initializes a bunch of the fields and closes the done() channel. However, every caller of `policyAddToQueue()` immediately calls `<-result.done`. If this is the case, then why does the caller need to know about the done channel at all? Couldn't `policyAddToQueue()` just block before returning, then return the fully-initialized context result?

The nice property about this is that it would make it less error-prone, as only one thread should be owning the `result` at a point in time. The callers of this function wouldn't be able to access partially-initialized values.

Reply:
I suppose a reason to return something like this is if we want to allow a non-blocking invocation. The current interaction with the DNS proxy happens to work but it's a little "just so" (dnsproxy -> NotifyOnDNSMsg -> UpdateGenerateRules -> policyAddToQueue, but this only happens after we return the DNS response to a lookup so blocking the proxy then is less bad).
Since it is simpler to make this always block maybe we can do that but add comments in the call-chain above so later refactors don't mess it up.

Reply:
@joestringer is right, this is absolutely racy right now. The pointer is returned and also queued into the channel and later accessed in parallel in the goroutine.

A much cleaner solution is to return a channel on which the caller can block on and the channel will return the result.